### PR TITLE
Fix HierarchicalPathFinder considering some unreachable cells as reachable

### PR DESCRIPTION
**Pull request by RoosterDragon**
When using the internal AbstractCellForLocalCell method to check if a local cell is reachable, this should return null when the cell is unreachable. If multiple abstract cells were required for that grid, this worked as intended. Only reachable cells are stored in the localCellToAbstractCell mapping. For a grid that required only a single abstract cell, which is the common case, we optimize this to store only the single abstract cell rather than the whole mapping for potentially 100 cells in that grid. However this makes no distinction between the reachable and unreachable cells, so when we check later we get incorrect results. If a cell is unreachable but belongs to the same grid as a single group of reachable cells then we incorrectly report it as reachable. The easiest way to see this incorrect behaviour is when the PathExists is called and can sometimes indicate a path exists when it does not.

To fix this, we now ensure we perform a check to see if the cell is reachable in this single layer case, this allows us to retain the optimization where we don't need to store the whole mapping, but allows us to correctly indicate when cells are unreachable.

I have confirmed this resolves the suspect behaviour noted in https://github.com/OpenRA/OpenRA/pull/20227#issuecomment-1233285720.

### CODE REVIEW

**Comment by PunkPun** [comment]
Need a rebase

**Comment by RoosterDragon** [comment]
Rebased.

**Review by PunkPun — APPROVED** [review]
Overall LGTM. I tried reproducing the issue but to no avail, what would be a reliable reproduction case?

**Comment by RoosterDragon** [comment]
My test was the following:

- On #20227 AI boats should not be able to attack land units on another island, but they do.
- With both #20227 and this PR merged as one, AI boats correctly ignore land units on another island.

**Comment by penev92** [comment]
Sounds like this should be marked as depending on #20227.

File: OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs [hunk 1]
```diff
@@ -146,12 +146,21 @@ public GridInfo(CPos?[] singleAbstractCellForLayer, Dictionary<CPos, CPos> local
...
+			public CPos? AbstractCellForLocalCell(CPos localCell, HierarchicalPathFinder hpf)
 			{
 				var abstractCell = singleAbstractCellForLayer[localCell.Layer];
 				if (abstractCell != null)
+				{
+					// All reachable cells in the grid are joined together so only a single abstract cell was needed,
+					// but there may be unreachable cells in the grid which we must exclude.
+					if (hpf != null && !hpf.CellIsAccessible(localCell))
+						return null;
 					return abstractCell;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs:160]:
```suggestion
						return null;

					return abstractCell;
```

Comment (Mailaender) [reply OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs:160]: Actually, never mind. Cramping everything together is the code style of this file.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/470815e20d51b174afbb5c105b2ef05b1c2db2ae...919f4922d7a3467ca75952825f5c6a5fd0f634b1)

**Pull request merged**